Add StellarGraph concrete type and concrete algorithms for subgraph and weakly connected components.

paul-tqh-nguyen opened the pull request:
This commit adds the concrete type `StellarGraph`. 

It also adds concrete algorithms for subgraph extraction and weakly connected components. 

Some misc. notes:
* The `StellarGraph` concrete implementation of `subgraph.extract_subgraph` can accept any iterable. I arbitrarily chose to make it take a Python set. It's not clear if this is ideal. We can avoid translation costs directly caused by `metagraph`.
* In the translators, I don't do a check for `has_pandas` when importing pandas because I check `has_stellargraph` and `Pandas` is a dependency of `StellarGraph`.
* When translating from a `NetworkXGraph` to a `StellarGraph`, it's not clear how precise we want our ints or floats to be. I arbitrarily chose 64bit. It's not clear what's best or standard or if we should let the caller specify this in some way.
* The of `assert_equal` class method of `StellarGraph` is checking everything in a Pythonic/NetworkX-like/hash-table-like sort of way using Python sets and Python dictionaries and such. I didn't see a way of avoiding this without sorting. It's not clear which if using arrays (better memory locality but possibly costly sorting) is better or worse than using hash-tables (worse memory locality but avoids sorting cost). 
* The [StellarGraph](https://stellargraph.readthedocs.io/en/stable/) library supports edge weights directly but does not support node weights directly. They do support node features and edge features, which are vectors associated with each edge and node. Edge features and edge weights are stored independently in the [StellarGraph](https://stellargraph.readthedocs.io/en/stable/) library. I supported node weights in the `StellarGraph` concrete type by using node features (using some position in the node feature vector specified at initialization time) but edge weights via their direct edge weight support. I figured this was worth pointing out because it might seem strange.

paul-tqh-nguyen committed:
Add scaffolding for metagraph-stellargraph

This commit adds the scaffollding for the metagraph-stellargraph plugin.

It currently has no algorithms, translators, etc.

paul-tqh-nguyen committed:
Add StellarGraph concrete type + subgraph and weakly connected components.

This commit adds the StellarGraph concrete type as well as the concrete algorithms for

This also includes StellarGraph <-> NetworkXGraph translators.

paul-tqh-nguyen committed:
Fix typo in metagraph_stellargraph/plugins/__init__.py

This commit fixes a copy-&-edit typo in metagraph_stellargraph/plugins/__init__.py .

paul-tqh-nguyen committed:
Remove debugging lines in metagraph_stellargraph/plugins/stellargraph/translators.py

This commit removes some debugging lines in metagraph_stellargraph/plugins/stellargraph/translators.py that are not strictly necessary.

paul-tqh-nguyen committed:
Make StellarGraph init checks more readable

This commit abstracts out a long expression for the valid index end value into a variable to make things more readable.

paul-tqh-nguyen committed:
Add embedding.train.node2vec for metagraph-stellargraph plugin

This commit adds the ndoe2vec algorithm for the StellarGraph plugin.

paul-tqh-nguyen commented:
https://github.com/metagraph-dev/metagraph-stellargraph/pull/1/commits/d84b992daa4e7cff602b39967b88deba47dfa28d adds the implementation of node2vec as well. 

NB: Working on this node2vec implementation made it clear that there needs to be an update to the vanilla `metagraph` test for node2vec. PR for that in the `metagraph` repo is forthcoming.

paul-tqh-nguyen committed:
Add concrete algorithm for `embedding.train.graphwave`

This commit adds a concrete implementation og GraphWave.

paul-tqh-nguyen commented:
https://github.com/metagraph-dev/metagraph-stellargraph/pull/1/commits/6134c32a0d63eaaf473b4f19b9d43132f07e2fc6 adds an implementation of GraphWave.

paul-tqh-nguyen committed:
Add StellarGraph implementation of GraphSAGE

This commit adds an implementation of GraphSAGE in the StellarGraph plugin. In particular, this commit adds util.graph_sage_node_embedding.apply and embedding.train.graph_sage.mean (uses the mean aggregator in GraphSAGE).

Some notes:
* When running GraphSAGE's training, this concrete algorithm checkpoints the modell at every epoch and uses the model with the smallest loss, not the model from the last epoch. The checkpointed models are deleted after training and will not exist after the concrete algorithm has completed running.
* Since a StellarGraph graph can't be modified after it has been initialized, sg_graph_sage_node_embedding_apply takes a NetworkX graph (since the initial node features are passed in via a NodeEmbedding (there's no clear better way to do this at the moment)) and creates a new StellarGraph from it. This seems expensive, but there's no clear better way of doing this. sg_graph_sage_mean_train does something similar.
* This commit adds the concrete type StellarGraphGraphSageNodeEmbedding as well since it is needed to implement GraphSAGE. Since StellarGraphGraphSageNodeEmbedding represents a neural network model whose weights are found via gradient descent (the weights from two training runs will almost never be the same), I left the assert_equal method as a stub.
* This commit also adds a test for GraphSAGE.

paul-tqh-nguyen commented:
https://github.com/metagraph-dev/metagraph-stellargraph/pull/1/commits/94176120b4db62da7176711bc2284a310072f5d2 adds an implementation of GraphSAGE in the StellarGraph plugin. In particular, this commit adds `util.graph_sage_node_embedding.apply` and `embedding.train.graph_sage.mean` (uses the mean aggregator in GraphSAGE).

Some notes:
* When running GraphSAGE's training, this concrete algorithm checkpoints the model at every epoch and uses the model with the smallest loss, not the model from the last epoch. The checkpointed models are deleted after training and will not exist after the concrete algorithm has completed running.
* Since a StellarGraph graph can't be modified after it has been initialized, `sg_graph_sage_node_embedding_apply` takes a NetworkX graph (since the initial node features are passed in via a NodeEmbedding (there's no clear better way to do this at the moment)) and creates a new StellarGraph from it. This seems expensive, but there's no clear better way of doing this. `sg_graph_sage_mean_train` does something similar.
* This commit adds the concrete type `StellarGraphGraphSageNodeEmbedding` as well since it is needed to implement GraphSAGE. Since `StellarGraphGraphSageNodeEmbedding` represents a neural network model whose weights are found via gradient descent (the weights from two training runs will almost never be the same), I left the `assert_equal` method as a stub.
* This commit also adds a `MultiVerify` test for GraphSAGE. It uses `scikit-learn` (which is fine since it's a dependency of StellarGraph). Since this `MultiVerify` test requires types not available in core Metagraph and available in the StellarGraph plugin, I included the test here rather than in core Metagraph.

paul-tqh-nguyen committed:
Add comments to GraphSAGE test

The GraphSAGE test is fairly simple and straight forward but the code is long since we have to generate the training and testing graphs and verify the results using K-NN.

This commit adds comments intended to make the code easier to read.

paul-tqh-nguyen commented:
https://github.com/metagraph-dev/metagraph-stellargraph/pull/1/commits/68d691206ad383375d91af0e7a934a40c4d232b2 adds code comments to the GraphSAGE test to make it easier to read.

jim22k commented on environment.yml:
```diff
@@ -0,0 +1,30 @@
+  - conda-forge::pre-commit
+  - setuptools
+
+# metagraph dependencies
```
I don't see `metagraph` listed here. Rather than duplicating the metagraph dependencies, you should simply list `metagraph::metagraph` (it's on the metagraph channel currently) as a dependency and many of the others become automatic dependencies as a result.

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/metagraph-stellargraph/pull/1/commits/da906041d7d47232ee0ea75327e8c3f2c435e498 implements this suggestion.

jim22k commented on metagraph_stellargraph/plugins/__init__.py:
```diff
@@ -0,0 +1,26 @@
+    from . import stellargraph
+
+    registry.register_from_modules(
+        stellargraph, name="metagraph_stellargraph_stellargraph"
```
Why do you give this a custom name? The PluginRegistry is already initialized with the default name "metagraph_stellargraph". For anyone calling an exact algorithm, they will want to call `resolver.plugins.metagraph_stellargraph.algos.subgraph.extract()` rather than `resolver.plugins.metagraph_stellargraph_stellargraph.algos.subgraph.extract()`

paul-tqh-nguyen replied:
The original idea was to match the convention used in `metagraph-cuda` of `<repo_name>-<library_name>`, but that idea doesn't really apply here. 

https://github.com/metagraph-dev/metagraph-stellargraph/pull/1/commits/8fa3546724399048c9910747c6164441a651ad9f implements this suggestion.

paul-tqh-nguyen committed:
Update environment.yml

This commit lists metagraph::metagraph as a dependency rather than enumerating all the dependencies of metagraph.

paul-tqh-nguyen committed:
Rename plugin from "metagraph_stellargraph_stellargraph" -> "metagraph_stellargraph"

This commit will make it so that calling algorithms directly by name is simpler.
